ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/CriterionIndicator.java

```diff
@@ -0,0 +1,124 @@
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package org.ta4j.core.indicators.helpers;
+
+import org.ta4j.core.AnalysisCriterion;
+import org.ta4j.core.BarSeries;
+import org.ta4j.core.BaseTradingRecord;
+import org.ta4j.core.Position;
+import org.ta4j.core.Trade;
+import org.ta4j.core.TradingRecord;
+import org.ta4j.core.indicators.CachedIndicator;
+import org.ta4j.core.num.Num;
+
+/**
+ * Criterion indicator.
+ * 
+ * <p>
+ * Transforms any AnalysisCriterion into an Indicator. Returns <code>true</code>
+ * if the calculated criterion value till a bar index is better than the
+ * {@link #requiredCriterionValue} or equal, otherwise returns
+ * <code>false</code>.
+ */
+public final class CriterionIndicator extends CachedIndicator<Boolean> {
+
+    private final AnalysisCriterion criterion;
+    private final Num requiredCriterionValue;
+    private final TradingRecord tradingRecord;
+    private final Position position;
+
+    /**
+     * Constructor.
+     * 
+     * @param series                 the bar series
+     * @param tradingRecord          the trading record
+     * @param criterion              the criterion to get the calculated criterion
+     *                               value on a bar index
+     * @param requiredCriterionValue the required criterion value to test against
+     *                               the calculated criterion value
+     */
+    public CriterionIndicator(BarSeries series, TradingRecord tradingRecord, AnalysisCriterion criterion,
+            Num requiredCriterionValue) {
+        super(series);
+        this.tradingRecord = tradingRecord;
+        this.criterion = criterion;
+        this.requiredCriterionValue = requiredCriterionValue;
+        this.position = null;
+    }
+
+    /**
+     * Constructor.
+     * 
+     * @param series                 the bar series
+     * @param position               the position
+     * @param criterion              the criterion to get the calculated criterion
+     *                               value on a bar index
+     * @param requiredCriterionValue the required criterion value to test against
+     *                               the calculated criterion value
+     */
+    public CriterionIndicator(BarSeries series, Position position, AnalysisCriterion criterion,
+            Num requiredCriterionValue) {
+        super(series);
+        this.tradingRecord = null;
+        this.criterion = criterion;
+        this.requiredCriterionValue = requiredCriterionValue;
+        this.position = position;
+    }
+
+    @Override
+    protected Boolean calculate(int index) {
+        if (tradingRecord != null && !tradingRecord.getPositions().isEmpty()) {
+            TradingRecord tradingRecordTillIndex = new BaseTradingRecord(tradingRecord.getStartingType(),
+                    tradingRecord.getTransactionCostModel(), tradingRecord.getHoldingCostModel());
+            for (Position pos : tradingRecord.getPositions()) {
```

Comment:
Another alternative to the current implementation would be to have a method within `TradingRecord` to get a shallow copy (objects within the trades-list are only copied by reference):

```
    @Override
    public TradingRecord getTradingRecord(int startIndex, int endIndex) {
        if (startIndex < 0) {
            throw new IllegalArgumentException(String.format("the startIndex: %s must not be negative", startIndex));
        }
        if (startIndex >= endIndex) {
            throw new IllegalArgumentException(
                    String.format("the endIndex: %s must be greater than startIndex: %s", endIndex, startIndex));
        }
        TradingRecord shallowClone = new BaseTradingRecord(startingType, transactionCostModel, holdingCostModel);
        if (!trades.isEmpty()) {
            shallowClone.getTrades().addAll(trades.subList(startIndex, endIndex));
        }
        return shallowClone;
    }
```
@MarcDahlem @team172011 what do you think? Is it worth the effort?